### PR TITLE
MSP Alarm Support

### PR DESCRIPTION
**Pull request by dustin**
This change creates a new MSP command that allows the peer (e.g. OSD) to ask for alarms from the flight controller. The alarm message includes severity and a message. In its current form, only one alarm is sent (fairly arbitrarily, other than severity taking precedence). Also, as a special and very useful case, boot reason is considered a warning from the perspective of the MSP code for the first few seconds. It might be worth generally considering that an alarm which would allow this special case code to be removed.

This has been very useful to me in showing both regular alarms (e.g. I let someone fly my sparky yesterday and saw 'MAN:NO RX' flashing on the screen because OpenTX companion turned off the radio for that profile) as well as boot reasons (saw watchdog firing in flight instead of mysteriously crashing).

STATUS: Flown a lot. Used to diagnose bugs.
CAVEAT: MWOSD side is not yet accepted (waiting for @ShikOfTheRa in ShikOfTheRa/scarab-osd#141)

### CODE REVIEW

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 1]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+#define ALARM_WARN 1
+#define ALARM_ERROR 2
+#define ALARM_CRIT 3
+
+static void msp_send_alarms(struct msp_bridge *m) {
+	union {
+		uint8_t buf[0];
+		struct {
+			uint8_t state;
+			uint8_t msg[MAX_ALARM_LEN];
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Should be char rather than uint8_t?

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 2]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+	SystemAlarmsData alarm;
+	SystemAlarmsGet(&alarm);
+
+	// Special case early boot times -- just report boot reason
+	if (PIOS_Thread_Systime() < BOOT_DISPLAY_TIME_MS) {
+		data.alarm.state = ALARM_CRIT;
+		strncpy((char*)data.alarm.msg,
+			(const char*)boot_reason_names[alarm.RebootCause],
+			sizeof(*boot_reason_names));
+		data.alarm.msg[sizeof(*boot_reason_names)] = 0;
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: I would prefer explicit null char here, `'\0'`, minor thing though.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 3]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+		msp_send(m, MSP_ALARMS, data.buf, strlen((char*)data.alarm.msg)+1);
+		return;
+	}
+
+	data.alarm.state = ALARM_OK;
+
+	for (int i = 0; i < SYSTEMALARMS_ALARM_NUMELEM; i++) {
+		if (alarm_names[i][0] != 0 &&
+		    ((alarm.Alarm[i] == SYSTEMALARMS_ALARM_WARNING) ||
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Mix of tabs and spaces...

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 4]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+
+			// Only take this alarm if it's worse than the previous one.
+			if (state > data.alarm.state) {
+				data.alarm.state = state;
+				switch (i) {
+				case SYSTEMALARMS_ALARM_SYSTEMCONFIGURATION:
+					strncpy((char*)data.alarm.msg,
+						(const char*)config_error_names[alarm.ConfigError],
+						sizeof(*config_error_names));
+					data.alarm.msg[sizeof(*config_error_names)] = 0;
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Explicit null char as above.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 5]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+				case SYSTEMALARMS_ALARM_MANUALCONTROL:
+					strncpy((char*)data.alarm.msg,
+						(const char*)manual_control_names[alarm.ManualControl],
+						sizeof(*manual_control_names));
+					data.alarm.msg[sizeof(*manual_control_names)] = 0;
+					break;
+				default:
+					strncpy((char*)data.alarm.msg,
+						(const char*)alarm_names[i], sizeof(*alarm_names));
+					data.alarm.msg[sizeof(*alarm_names)] = 0;
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Explicit null char as above.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 6]
```diff
@@ -391,6 +397,154 @@ static void msp_send_boxids(struct msp_bridge *m) {
...
+					strncpy((char*)data.alarm.msg,
+						(const char*)config_error_names[alarm.ConfigError],
+						sizeof(*config_error_names));
+					data.alarm.msg[sizeof(*config_error_names)] = 0;
+					break;
+				case SYSTEMALARMS_ALARM_MANUALCONTROL:
+					strncpy((char*)data.alarm.msg,
+						(const char*)manual_control_names[alarm.ManualControl],
+						sizeof(*manual_control_names));
+					data.alarm.msg[sizeof(*manual_control_names)] = 0;
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Explicit null char as above.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 1 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Sure.  :)

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 2 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: I never use `'\0'`, but I'll do that for you.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 3 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: There seems to be no other way to achieve sensible indentation here. :/

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 4 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Done

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 5 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Done

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 6 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Done.

**Comment by mlyle** [comment]
jenkins, test this please

**Comment by mlyle** [comment]
jenkins, test this please

**Comment by mlyle** [comment]
```/var/lib/jenkins/jobs/dron/workspace@2/Nodes/master/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:510:9: error: 'SYSTEMALARMS_ALAfRM_WARNING' undeclared (first use in this function)
    case SYSTEMALARMS_ALAfRM_WARNING:
```

**Comment by dustin** [comment]
Sorry, PicoC incompatibility.  Fixed.

**Comment by mlyle** [comment]
;)

**Pull request merged**